test/e2e/k8s_util.go

```diff
@@ -99,34 +100,35 @@ func (k *KubernetesUtils) probe(
 	cmd := []string{
 		"/bin/sh",
 		"-c",
+		fmt.Sprintf("for i in $(seq 1 3); do /agnhost connect %s:%d --timeout=1s --protocol=%s; done;", dstAddr, port, strings.ToLower(string(protocol))),
 	}
-	switch protocol {
-	case v1.ProtocolTCP:
-		cmd = append(cmd, fmt.Sprintf("for i in $(seq 1 3); do /agnhost connect %s:%d --timeout=1s --protocol=tcp && exit 0 || true; done; exit 1", dstAddr, port))
-	case v1.ProtocolUDP:
-		cmd = append(cmd, fmt.Sprintf("for i in $(seq 1 3); do /agnhost connect %s:%d --timeout=1s --protocol=udp && exit 0 || true; done; exit 1", dstAddr, port))
-	case v1.ProtocolSCTP:
-		cmd = append(cmd, fmt.Sprintf("for i in $(seq 1 3); do /agnhost connect %s:%d --timeout=1s --protocol=sctp && exit 0 || true; done; exit 1", dstAddr, port))
-	}
-	// HACK: inferring container name as c80, c81, etc, for simplicity.
-	containerName := fmt.Sprintf("c%v", port)
 	log.Tracef("Running: kubectl exec %s -c %s -n %s -- %s", pod.Name, containerName, pod.Namespace, strings.Join(cmd, " "))
 	stdout, stderr, err := k.runCommandFromPod(pod.Namespace, pod.Name, containerName, cmd)
-	var curConnectivity PodConnectivityMark
 	if err != nil {
 		// log this error as trace since may be an expected failure
 		log.Tracef("%s -> %s: error when running command: err - %v /// stdout - %s /// stderr - %s", podName, dstName, err, stdout, stderr)
 		// do not return an error
-		if strings.Contains(stderr, "TIMEOUT") {
-			curConnectivity = Dropped
-		} else {
-			curConnectivity = Rejected
-		}
-	} else {
-		curConnectivity = Connected
+		return decideProbeResult(stderr, 3)
 	}
+	return Connected
+}
+
+// decideProbeResult uses the probe stderr to decide the connectivity.
+func decideProbeResult(stderr string, probeNum int) PodConnectivityMark {
+	countConnected := probeNum - strings.Count(stderr, "\n")
+	countDropped := strings.Count(stderr, "TIMEOUT")
+	countRejected := strings.Count(stderr, "REFUSED") + strings.Count(stderr, "no route to host")
 
-	return curConnectivity
+	if countRejected == 0 && countConnected > 0 {
+		return Connected
+	}
+	if countConnected == 0 && countRejected > 0 {
+		return Rejected
+	}
+	if countDropped == probeNum {
+		return Dropped
+	}
+	return Error
```

Comment:
wasn't this part of another PR aimed at fixing FQDN e2e test flakes?

Reply:
Yes, this is in that FQDN flake PR. Because this PR is also kind of involved in this `decideProbeResult` issue, I'm thinking if this could be merged by this PR?